tools/server-spring/src/main/java/org/eclipse/rdf4j/http/server/repository/transaction/ActiveTransactionRegistry.java

```diff
@@ -94,24 +105,59 @@ private ActiveTransactionRegistry() {
 
 		secondaryCache = CacheBuilder.newBuilder()
 				.removalListener((RemovalNotification<UUID, Transaction> notification) -> {
+					logger.debug("secondary cache removal");
 					if (RemovalCause.EXPIRED.equals(notification.getCause())) {
 						final UUID transactionId = notification.getKey();
 						final Transaction entry = notification.getValue();
+						logger.debug("expired transaction to be removed {}", transactionId);
 						synchronized (primaryCache) {
-							if (!entry.hasActiveOperations()) {
-								// no operation active, we can decommission this entry
-								primaryCache.invalidate(transactionId);
-								logger.warn("deregistered expired transaction {}", transactionId);
-							} else {
-								// operation still active. Reinsert in secondary cache.
-								getSecondaryCache().put(transactionId, entry);
+							// no operation active, we can decommission this entry
+							primaryCache.invalidate(transactionId);
```

Comment:
The contract (as I understand it) is that a cliend needs to ping the transaction endpoint before the timeout expires, if it do not do that the secondsry cache is not updated which means that the whole transaction timeout and need to be invalidated. Id th eclient do ping the endpoint, the seconday cache is upate and the entry stays there (e.g. the removal listener is not invoked)

Reply:
Thanks, I think that's correct. I was worrying about possible operations on the transaction operation executor being left "dangling" so to speak, but on reviewing I think the logic in `Transaction#close` takes care of that.

Reply:
(I am rusty on how this registry works exactly so it's entirely possible I've read the implications wrong)

Reply:
The remote transaction is always explicitely commited by sending a request to the TransactionController. If you are concerned that it may be interrupted by closing the connection when the entry expired from the secondary cache, I could add a synchronized section/mutex (over the primaryCache) there so to prevent that from happening. all the code that deals with canceling stale transactions witin ActiveTransactionRegisty does it within synchronised sections over the primatCache ...

Reply:
Thanks @damyan-ognyanov , I 've taken another look at how this is supposed to work and I think I understand a little better now.

For some reason I had it in my head that we dealt with transaction operations asynchronously, in which case it would be vital to keep a server-side transaction active as long as there are scheduled tasks for it - even if the client has already sent a commit and has stopped doing pings.

But of course, that is not actually how this works: while we do use asynchronous processing to some extent, each transaction operation `Future` waits for completion until it sends a response back to the client - so it's all still synchronous in the end. 

It also means that what I said above, about the client sending mulitple operations, cannot actually happen in practice: the client would be blocked from sending the next operation until the first one has completed. 

So with all that in mind I think simplifying this is fine - as you say the one corner case we need to take care of is when a commit operation takes a long time and the client stops sending pings halfway through. And even there I'm no longer convinced we need to do something extra, to be honest.

Reply:
Thank's @jeenbroekstra, I also think that overcomlicating the handling of the corner case when the transaction is in commit but expires from secondary cache altough it may solves it (by using an extra flag `inCommit` within `Transaction.commit() that then is checked within the removal listner to decide whether to keep or discard) that also can introduce a means to continue keeping a (possibly) stale transaction indefenitely, in case the transaction is blocked within the underlying sail during the handling of `commit()` ...

Let me know if you need something else to be done for this PR.